### PR TITLE
feat(dao) support different read and write consistency levels in cassandra

### PR DESCRIPTION
**Pull request by Abhishekvrshny**
### Summary

This adds the capability to define different read and write
consitency levels for cassandra. Earlier both of these were set
by a single `cassandra_consistency` configuration parameter.

This PR introduces `cassandra_write_consistency` and
`cassandra_read_consistency` configuration parameters and removes
`cassandra_consistency` parameter.

### CODE REVIEW

**Comment by bungle** [comment]
Hi @Abhishekvrshny, nice PR. Should we still keep `cassandra_consistency` for backward compatibility. E.g. if a more specific consistency is not defined, use that?

**Comment by Abhishekvrshny** [comment]
@bungle Sure. In that case, should the PR be against `master` or `next`? Also, is there a way to mark a config as deprecated somehow, for users to migrate?
And I see some PR checks failing, not sure how to resolve those.

**Comment by Abhishekvrshny** [comment]
@bungle I have made the suggested changes to the best of my understanding. Please see if this works.

**Comment by Abhishekvrshny** [comment]
@bungle Checking if you got a chance to review this.

File: kong.conf.default [hunk 1]
```diff
@@ -821,9 +821,23 @@
...
+#cassandra_consistency = DISABLED     # Consistency setting to use when reading/
+                                      # writing to the Cassandra cluster.
+                                      # Note: `cassandra_write_consistency` and
+                                      # `cassandra_read_consistency` should be
+                                      # preferred over this, with this `DISABLED`
+                                      # If this is `DISABLED`, `cassandra_write_consistency`
+                                      # and `cassandra_read_consistency` would be used,
+                                      # else `cassandra_consistency` takes precedence over
+                                      # `cassandra_write_consistency` and `cassandra_read_consistency`
+                                      # It exists only for backward compatibility.
```
Comment (thibaultcha) [inline kong.conf.default:833]: We do not leave deprecated properties in `kong.conf.default`, this property can/should be removed from here.

File: kong/templates/kong_defaults.lua [hunk 2]
```diff
@@ -115,7 +115,9 @@ cassandra_ssl = off
 cassandra_ssl_verify = off
 cassandra_username = kong
 cassandra_password = NONE
-cassandra_consistency = ONE
+cassandra_consistency = DISABLED
```
Comment (thibaultcha) [inline kong/templates/kong_defaults.lua:118]: Should be set to `NONE` instead?

File: kong/conf_loader.lua [hunk 3]
```diff
@@ -336,6 +336,38 @@ local CONF_INFERENCES = {
...
+     "LOCAL_QUORUM",
+     "ONE",
+     "TWO",
+     "THREE",
+     "LOCAL_ONE",
+    },
+    deprecated = {
+      replacement = "cassandra_write_consistency / cassandra_read_consistency",
+      alias = function(conf)
+        if conf.cassandra_consistency ~= "DISABLED" then
```
Comment (thibaultcha) [inline kong/conf_loader.lua:352]: I don't think that we need this new enum value. By replacing the default value to `NONE`, we already prevent new instances of the config from loading it (therefore only r/w consistencies are loaded). Then, if `cassandra_consistency` is specified, we overwrite the r/w consistencies in this aliasing. The warning log will instruct users that the property is deprecated and that they should favor the new r/w properties instead.

**Comment by Abhishekvrshny** [comment]
@thibaultcha Thanks for the review. I have made the required changes.

File: spec/01-unit/03-conf_loader_spec.lua [hunk 4]
```diff
@@ -516,13 +516,41 @@ describe("Configuration loader", function()
...
+      }))
+      assert.equal("TWO", conf.cassandra_read_consistency)
+      assert.equal("QUORUM", conf.cassandra_write_consistency)
+
+      conf = assert(conf_loader(nil, {
+        cassandra_consistency = "QUORUM",
+        cassandra_write_consistency = "LOCAL_QUORUM"
+      }))
+      assert.equal("QUORUM", conf.cassandra_read_consistency)
+      assert.equal("LOCAL_QUORUM", conf.cassandra_write_consistency)
```
Comment (thibaultcha) [inline spec/01-unit/03-conf_loader_spec.lua:552]: Please add new tests around the deprecation of properties to a new `describe` block instead. Leave this test unchanged, and add a new `describe` block to this test suite. I did the same in #5771 which isn't merged yet, so you can just add the new `describe` block yourself, in the same place as I added it in #5054: 

```lua
describe("deprecated properties", function()
  it("cassandra_consistency -> cassandra_r/w_consistency", function()

  end)
end)
```

File: kong/conf_loader.lua (hunk 3 above)
Comment (Abhishekvrshny) [reply kong/conf_loader.lua]: @thibaultcha done

File: spec/01-unit/03-conf_loader_spec.lua (hunk 4 above)
Comment (Abhishekvrshny) [reply spec/01-unit/03-conf_loader_spec.lua]: @thibaultcha done

**Comment by Abhishekvrshny** [comment]
@thibaultcha gentle reminder for reviewing the last change. Thanks.

**Comment by thibaultcha** [comment]
Merged to next with some edits, thank you!

**Comment by hbagdi** [comment]
@Abhishekvrshny Please fill out the following form to claim your contributor swag: https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributor-t-shirt

**Pull request closed without merging**